### PR TITLE
Fix index ops on expanded tensors in Inductor

### LINKED ISSUE #183986: `torch.compile` silently produces wrong results for `index_add`/`index_copy`/`index_fill`/`index_put` on expanded (stride-0) tensors
### 🐛 Describe the bug

When a tensor is created via `.expand()` (introducing stride-0 dimensions), `index_add`, `index_copy`, `index_fill`, and `index_put` under `torch.compile(backend='inductor')` produce **silently wrong forward results**. The compiled code operates directly on the aliased storage instead of materializing the expansion, so writes intended for independent rows collide through the shared underlying memory.

Eager mode and `aot_eager` are correct. Only the Inductor backend is affected.

### Reproducer

```python
import torch

def fn(src, idx):
    x = torch.zeros(1, 8, device="cuda").expand(4, -1)  # stride-0 in dim 0
    return x.index_add(0, idx, src)

src = torch.ones(3, 8, device="cuda")
idx = torch.tensor([0, 1, 0], device="cuda")

eager = fn(src, idx)
compiled = torch.compile(fn, backend="inductor")(src, idx)

print("Eager row 0:", eager[0].tolist())      # [2, 2, 2, 2, 2, 2, 2, 2]
print("Compiled row 0:", compiled[0].tolist()) # [3, 3, 3, 3, 3, 3, 3, 3]  ← WRONG
print("Eager row 2:", eager[2].tolist())       # [0, 0, 0, 0, 0, 0, 0, 0]
print("Compiled row 2:", compiled[2].tolist()) # [3, 3, 3, 3, 3, 3, 3, 3]  ← WRONG
```

**Expected:** Row 0 = `[2, ...]` (idx 0 hit twice), Row 1 = `[1, ...]`, Rows 2–3 = `[0, ...]`
**Actual (compiled):** All rows = `[3, ...]` — the total sum leaks to every row via aliased storage.

### All affected ops

```python
import torch

def test_op(name, op_fn):
    torch._dynamo.reset()
    eager = op_fn()
    compiled = torch.compile(op_fn, backend="inductor")()
    diff = (eager - compiled).abs().max().item()
    print(f"  {name}: max_diff={diff:.4f}  {'BUG' if diff > 0.01 else 'OK'}")

idx = torch.tensor([0, 1, 0], device="cuda")
src = torch.ones(3, 8, device="cuda")

test_op("index_add", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_add(0, idx, src))
test_op("index_copy", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_copy(0, idx[:2], src[:2]))
test_op("index_fill", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_fill(0, idx[:2], 1.0))
test_op("index_put(acc=True)", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_put((idx,), src, accumulate=True))
test_op("index_put(acc=False)", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_put((idx[:2],), src[:2], accumulate=False))
# All print BUG
```

`scatter`, `scatter_add`, `scatter_reduce`, `select_scatter`, `slice_scatter` are NOT affected — they go through a different codegen path that correctly handles stride-0 inputs.

### Controls

| Condition | Result |
|-----------|--------|
| Non-expanded tensor (`torch.zeros(4, 8)`) | Correct (diff=0) |
| `.clone()` after `.expand()` | Correct (diff=0) |
| `aot_eager` backend | Correct (diff=0) |
| All shapes tested (4×8 to 128×256) | All wrong |
| All dtypes (float16, float32, float64) | All wrong |
| Expand on dim 0 or dim 1 | Both wrong |
| Backward (src.grad) | Correct — only forward is affected |

### Root cause

Inductor's lowering for the `index_*` family does not detect or materialize expanded (stride-0) tensors before performing indexed writes. Because all expanded rows alias the same storage, writes to different logical rows collide.

### Workaround

Insert `.clone()` or `.contiguous()` after `.expand()`:
```python
x = torch.zeros(1, 8, device="cuda").expand(4, -1).clone()  # materialize
```

### Relationship to existing issues

This is the same bug class as #180164 / #182012 (stride-0 tensor mishandling in Inductor), but:
- #180164 (`slice_scatter`/`select_scatter` backward): **fixed**
- #182012 (`diagonal_scatter` backward): **open, fix PR ready**
- This bug: `index_add`/`index_copy`/`index_fill`/`index_put` **forward** — NOT fixed

The `index_*` ops are a distinct code path from the `*_scatter` ops, so existing fixes do not cover them.

### Versions

- PyTorch: 2.13.0.dev20260513+cu126
- GPU: Tesla T4 (SM 7.5), also reproducible on CPU
- OS: Linux

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by PannenetsF**
Fixes #183986.

## What

This fixes silent wrong results for `torch.compile(..., backend="inductor")` when `index_add`, `index_copy`, `index_fill`, or `index_put` writes into an expanded tensor with internal overlap, such as a stride-0 dimension from `.expand()`.

## Why

Inductor's reinplace pass can turn functional indexed writes into in-place indexed writes when it decides the mutation is not observable. That is unsafe for expanded tensors with internal overlap: distinct logical indices can alias the same physical storage, so writes that should target different logical rows collide.

For example, writing into `torch.zeros(1, 8).expand(4, -1)` can incorrectly accumulate into the shared backing row and then expose that value through every expanded row.

`index_fill` also tried to preserve the input strides after lowering through `index_copy`. For overlapping expanded inputs, eager returns a materialized non-overlapping result instead, so preserving the stride-0 layout is not correct.

## How

The fix has two parts:

1. Prevent the Inductor reinplace pass from reinplacing an operation when the candidate mutated argument has definite internal overlap.
2. Update the `index_fill` reference decomposition so it only preserves strides for non-overlapping inputs.

Regression tests cover expanded tensor destinations for:
- `index_add`
- `index_copy`
- `index_fill`
- `index_put(..., accumulate=True)`
- `index_put(..., accumulate=False)`

They also include an `index_add` case indexing along a non-expanded dimension while the destination still contains a stride-0 expanded dimension.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
fix merge conflicts

**Comment by PannenetsF** [comment]
> fix merge conflicts

hi @jansel , i have fixed it.

**Review by jansel — DISMISSED** [review]

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Review by jansel — CHANGES_REQUESTED** [review]
fix test failures

**Review by jansel — DISMISSED** [review]

**Comment by PannenetsF** [comment]
Hi @jansel, I looked into the revert a bit more. The failure seems specific to the Pallas CPU backend rather than the reinplace correctness fix. 

My current suspicion is that after the expanded tensor is materialized, Pallas lowers the indexed write through JAX scatter with a wrong update/broadcast shape. The failure reported a shape mismatch like `(1, 2, 8)` being broadcast to `(1, 2, 1, 8)`, which points to the Pallas scatter/index_copy lowering rather than the overlap guard itself.

For this reland I marked the two new expanded-indexing tests as expected failures for Pallas using the existing sentinel mechanism, while keeping the tests active for the regular Inductor backends. Does this direction look okay to you, or would you prefer that I split out / investigate the Pallas backend issue first?

**Comment by jansel** [comment]
Skipping failing pallas tests is fine, but looks like you have CI failures (fix or rebase).

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by PannenetsF** [comment]
@jansel Hi, seems the failures are not related to this PR and I have rebased multi-times. Shall we skip those test and just merge the PR?

**Comment by jansel** [comment]
Make sure to rebase to the `viable/strict` branch.  I'd like to see a green CI before merging.

@pytorchbot rebase

**Comment by PannenetsF** [comment]
hi @jansel , could you also review this when possible? thanks!

**Comment by jansel** [comment]
@claude are those CI failures related?  How confident are you that the reasons for the revert have been addressed?  Is this safe to reland?

**Review by jansel — DISMISSED** [review]

**Comment by PannenetsF** [comment]
Hi @atalman, could you share what internal failure or regression triggered the revert? thanks!

**Comment by PannenetsF** [comment]
Hi @jansel , could you review this again? not sure what led to the revert. thanks!

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
This should be safe to reland, the failing test was updated in D107445387 --- more details for meta folks: https://metamate.internalmeta.com/artifact/late-mustang-13

**Comment by jansel** [comment]
@claude are those CI failures related to this PR?
